### ISSUE TITLE
Restoring a session with buffers opened in specific tabs opens all buffers in the first tab

### ISSUE THREAD

**Issue by lougreenwood** [AstroNvim/AstroNvim#1775]
### Checklist

- [X] Using a stable version of Neovim (i.e. not neovim nightly)
- [X] `:AstroUpdate`
- [X] Restarted AstroNvim

### Operating system/version

MacOS 13.3.1

### Terminal/GUI

iTerms

### AstroNvim Health

2   astronvim: require("astronvim.health").check()
   3
   4   AstroNvim
   5   - AstroNvim Version: v3.10.3
   6   - Neovim Version: v0.9.0
   7   - OK Using stable Neovim >= 0.8.0
   8   - OK git is installed: Used for core functionality such as updater and plugin management
   9   - OK open is installed: Used for gx mapping for opening files with system opener (Optional)
  10   - OK lazygit is installed: Used for mappings to pull up git TUI (Optional)
  11   - OK node is installed: Used for mappings to pull up node REPL (Optional)
  12   - OK gdu is installed: Used for mappings to pull up disk usage analyzer (Optional)
  13   - WARNING btm is not installed: Used for mappings to pull up system monitor (Optional)
  14   - OK python is installed: Used for mappings to pull up python REPL (Optional)

### Describe the bug

If I save a session which has multiple tabs, where each tab shows a single buffer, when I re-open astronvim and restore that session all of the tabs will be restored, the first tab will show all of the open buffers in the tab bar and the remaining tabs will just show the expected buffers.

I already opened [an issue](https://github.com/Shatur/neovim-session-manager/issues/79) in `neovim-session-manager` the maintainer said they don't get this behaviour on 0.9, so that narrows it down my config or Astronvim defaults.

I also tried removing my config and opening Astronvim without any custom config and the issue persists.

Here's a video example:
https://monosnap.com/file/UHkuF3vVjjxYhSE0BHL78qDU0ktE5g

### Steps to Reproduce

1. Open an empty session
2. Open a file
3. Open another file into a new tab
4. Open a third file into a new tab
5. Save the session
6. See that tab #1 only shows one buffer with for the expected file
7. Close & re-open astronvim
8. Load the saved session
9. See that tab #1 now shows all of the (non-hidden) buffers opened in the session in it's buffer bar
10. See that the other tabs show their expected buffers

### Expected behavior

Restoring a session restores buffers to the tab which they were opened in and the first tab does not show all of the buffers in the session.

**Comment by mehalter** [comment]
Ah thanks for opening this! I think I know what the problem is here and I should have some time this upcoming week to look into it. It's probably related to the session management not restoring our variables that manage the buffers and their tab scoping

**Comment by mehalter** [comment]
Hm, this is an unfortunate issue that we have run into 😅 I think this might force our hand to move to `resession.nvim` for session management and actually replace the use of `mksession` (https://github.com/stevearc/resession.nvim/)

I think I can implement this where it will play nicely with our scoped tab setup relatively nicely and do it as an `opt-in` setting to avoid breaking changes for now. Then I can leave a work around here which would be enabling `resession` rather than `neovim-session-manager` until v4 release (whenever that is) of AstroNvim.

What do you think of that @lougreenwood ?

**Comment by lougreenwood** [comment]
@mehalter Not sure I fully grasp, so I'll re-itterate what I think you suggested just to make sure I understood:

- Seems astronvim needs to move to `resession.nvm` because "reasons" 😉 
- `resession.nvim` is a breaking change, so migrating now isn't possible
- instead you'll implement `resession.nvim` as an opt-in until V4
- when V4 us released, we can remove the opt-in

Is that correct?

Also, I didn't realise that we have tab scoping, so that means that things like [`scope.nvim`](https://www.google.com/search?client=safari&rls=en&q=scope.nvim&ie=UTF-8&oe=UTF-8) or [`tab-scope.nvim`](https://github.com/backdround/tabscope.nvim) aren't necessary with Astronvim? Out of curiosity, if that's true, do we have a util that can find the tab(s) that a specific buffer is in and return the tab? I'm wanting to implement a Telescope mapping to jump to a file if it exists, else open in a new tab and tab-scoped buffer utils would be useful for that.

Also, one more thing - if you decide to go with `resession.nvim`, please please please implement git branch-specific session saving 🤞 🙏 😆

**Comment by lougreenwood** [comment]
@mehalter I remembered that I had another question about astronvim's tab-scoped buffers.

Using the current api, do you think it's possible to `bdelete` a buffer when it's last split in the current tab or the current tab is closed (assuming there are no other instances in the).

Working primarily with tabs (like I do) and the tab-scoped buffering means that closing the last instance of a buffer in a tab strongly implies that I don't need that buffer around anymore.

WDYT? I can create another issue if you want to discuss this elsewhere?

**Comment by mehalter** [comment]
That's definitely possible @lougreenwood . You could make an autocmd that runs on our `AstroBufsUpdated` user event that then checks if there are any buffers not in a `vim.t.bufs` table, looping over all tab pages and then removing the ones that don't appear. It's definitely not the most performant approach but is possible. We don't do much fancy logic, but the tables are available and we have auto command events that let you hook in and add stuff that you want

**Comment by mehalter** [comment]
Also I have a start to this resession integration on `nightly` with a global option to opt into it. It seems to be working fully with multiple-tab session restoration. There's on bug upstream that I'm planning on making a PR to fix

**Comment by mehalter** [comment]
Also your original question about scope.nvim and similar plugins, those shouldn't be necessary in our configuration since we do our own tab management that handles scoping the tabs

**Comment by mehalter** [comment]
It's more like how we manage the buffer bar that scopes the tabs rather than their approach which marks tabs as unlisted. We just manage a tab local variable `vim.t.bufs` as buffers are created and deleted and then use those variables to render the tabline

**Comment by mehalter** [comment]
With resession I think we should also definitely be able to easily support git branch based session management

**Comment by mehalter** [comment]
@lougreenwood I have made a tag of release v3.11.0 which has the ability to use `vim.g.resession_enabled = true` to enable the experimental custom session management that allows for this tab restoration. This is pretty experimental and I can't fully promise stability. But if you want to give it a try and let me know that would be aweesome!

I am going to go ahead and close this with a workaround label added. If you have new findings feel free to continue discussion here.

**Comment by lougreenwood** [comment]
Thanks 🎉 ! I added the following to my `polish.lua`, but it doesn't seem to be working:

```lua
return function() {
  -- Enable experimental session support
  -- https://github.com/AstroNvim/AstroNvim/issues/1775#issuecomment-1520271952
  vim.g.resession_enabled = true
}
```

But I don't see any changes to functionality, seems I still have the same bug and also the sessions picker still seems to be showing the sessions I had from the old system, is this to be expected?

Maybe I'm setting things up wrong?

**Comment by lougreenwood** [comment]
I also see that `recession.nvim` is listed in `clean` & `disabled` when I check in lazy, do I also need to install & setup resession in my config?

**Comment by mehalter** [comment]
@lougreenwood you should use the `options` table to set `g` variables. That makes sure the option is set before the plugins are set up

**Comment by mehalter** [comment]
Here is an example: https://github.com/AstroNvim/user_example/blob/main/options.lua

**Comment by teddywaweru** [comment]
I've done some testing around this, & noticed some bugs on:
1. When closing a buffer in one tab, a buffer from another tab emerges in the current tab.
2. On saving multiple tabs as a single session, on restarting Neovim, the tab numbering that appears on the top right defaults back to the Neovim's tab layout.
![2023-07-12-024352_1920x1980_scrot](https://github.com/AstroNvim/AstroNvim/assets/57697272/5a74e59c-9f66-43b1-b8c2-217de579ff93)

**Comment by mehalter** [comment]
@teddywaweru for point 1 this is actually a bug in [mini.bufremove](https://github.com/echasnovski/mini.bufremove) It would be great to set up a minimal configuration that just has `mini.bufremove` and report that bug because that is pretty bad.

Also for point 2, I can't replicate what you are describing. It looks like your heirline is crashing for some reason which is causing it to not be there anymore. Here is a demo of me saving and restoring a multi-tab session: https://asciinema.org/a/466ZDKoiUYMXFm6okOqh0C8BT

**Comment by mehalter** [comment]
also here is another example doing `:qall` and then restoring the last session with `<leader>Sl` with multiple tabs: https://asciinema.org/a/n08Zca19WqXnPGDekOY3e0QNP

**Comment by teddywaweru** [comment]
May have been a fluke on my side on No. 2,, I considered that maybe the issue was enabling [autohiding the tabline](https://astronvim.com/Configuration/autocmds#example-autocommand-usage), so I disabled it and re-enabled, and now it works quite okay. I'll keep checking still.

For 1, I've made an unnecessarily long video for it, where I open buffers on different tabs, and start closing them in tab 1, but buffers on other tabs shift to tab 1, until none is left: https://asciinema.org/a/ysPvGAqaHj0OhHcnIdKybhvxi 
Buffers don't seem to respect tab allocation, and this is also applicable when using telescope's find buffer: it opens the buffer in the current open tab, while I'd say the expected behaviour is the window navigates/opens the buffer's tab...

**Comment by mehalter** [comment]
@teddywaweru Yeah definitely. I totally agree with your points on the buffers and tab allocation. I can make a `mini.bufremove` issue tomorrow or sometime to report this upstream since the bug is coming from there. For example if you use just `:bdelete` this doesn't happen.

**Comment by teddywaweru** [comment]
`bdelete` deletes the buffer's entire tab, yikes,,is that really the viable solution?
Also, would `bufremove` also be responsible for opening the `telescope find buffer` in the current tab? I feel that the concept of real tabs introduces significant changes here. On another case involving Neotree, if I have i the menu toggled on a tab, it takes three tries to enable it on a new tab.
I feel I may have hijacked this issue thread inappropriately.

**Comment by mehalter** [comment]
Haha no worries. The default behavior of Neovim/Vim is to delete the tab if you close the last buffer in the tab. So that would be the desired behavior that we want to replicate. So this is simply just a bug with `mini.bufremove`. As for the `neo-tree` issue that also is probably just more side effects of not many of these developers using real vim tabs and just more bugs in upstream plugins that need to be reported and fixed. I think people nowadays tend to not really user real vim tabs and so lots of these plugins just don't keep it in mind especially when they do some sort of buffer creation/deletion

**Comment by mehalter** [comment]
Also no, `mini.bufremove` doesn't have any role in opening files with telescope. My guess is if you have problems with it, it's almost certainly bugs with the interaction between Telescope.nvim and real vim tabs 😂 sorry this isn't much better news :/

**Comment by teddywaweru** [comment]
I have the #2126 pull request pending review which should address the closing of buffers. I think it's a proper fix. Kudos on the global `vim.t.bufs` variable; that really bypasses the concept of buffers not being associated with specific tabpages.
I also did sth on the Neotree issue I mentioned above, but I haven't been able to get any confirmation whether this is a personally isolated issue. Could you confirm? Case: if Neotree is opened in a tabpage, it takes about three <leader>e calls to get it to open it in a new tabpage...

**Comment by mehalter** [comment]
@teddywaweru I have added a bit of a hacky somewhat fix for this behavior in a slightly different way (https://github.com/AstroNvim/AstroNvim/pull/2121/commits/ecbdb5579e6dfb4f74b775b9efa893814edb1ae0) I have also opened the upstream issue. I will remove the hacky fix when the bug is fixed upstream.

**Comment by mehalter** [comment]
Also regarding the other bug you found in neo-tree. I can replicate that as well in a minimal neovim instance with just neo-tree isntalled. That would also be a good idea to report upstream.

**Comment by mehalter** [comment]
@teddywaweru also just so you know, I opened that issue upststream to `mini.bufremove` and it was closed as not planned and intended behavior: https://github.com/echasnovski/mini.nvim/issues/419

**Comment by teddywaweru** [comment]
yeah, I've seen the hack, but is it the expected behavior? `tabpagebuflist()` gives a list of windows in the tab, and not the buffers that are associated in the tab. That's why I had resorted to `vim.t.bufs` since this assures that we're dealing with buffers located only in the current tab.



> Also regarding the other bug you found in neo-tree. I can replicate that as well in a minimal neovim instance with just neo-tree isntalled. That would also be a good idea to report upstream.

I'll raise it. Hoepfully it's a better outcome than mini.nvim, though I have some hack that toggles the Neotree window when switching tabs....

For mini.nvim, would it be viable to have an empty buffer open rather than picking another from the jumplist? That would still maintain the layout, but I wouldn't know if the plugin has such a functionality, or whether that qualifies as a layout change too.

**Comment by echasnovski** [comment]
No, using a buffer which was not previously opened in the tabpage is not a bug in 'mini.nvim'. Tabpage is a collection of windows which can be used to view any buffers. There is simply no such (built-in) thing as "buffer binded to tabpage": you **can** show same buffer in multiple tabpages.

> For mini.nvim, would it be viable to have an empty buffer open rather than picking another from the jumplist?

This is be possible, but will not be a part of 'mini.bufremove' because it mostly violates the (Neo)Vim's views on how tabpage/windows/buffers should interact (which I find quite reasonable).

**Comment by mehalter** [comment]
@teddywaweru let me know what you think of the latest AstroNvim `nightly`, I think I finally found a pretty elegant solution to this :)

**Comment by teddywaweru** [comment]
I agree, it's a nice way to get it done, works on my system as expected.
Since this is progressing, what's your thought on how to approach browsing buffers? The telescope issue I mentioned before... Currently, the buffer opens in the current tab; would it be possible to navigate  to the buffer's tab?

**Comment by mehalter** [comment]
That would be a good question for the telescope people probably 😂 since we don't implement that functionality in any way

**Comment by mehalter** [comment]
Oh @teddywaweru you want to use `:drop` which telescope apparently does support

**Comment by mehalter** [comment]
Here's an example: https://github.com/nvim-telescope/telescope.nvim/issues/2188#issuecomment-1516861881

**Comment by teddywaweru** [comment]
That is such a satisfying solution! Love it!
And Neotree implements a similar `open_tab_drop` option too.
I think that's all I needed to really work with tabs fully. I'll report any niche cases that I'll run into. Thanks!
